Replace coreos-assembler ImageStream by new knobs in the pipecfg

jlebon opened the pull request:
See individual commits.

jlebon commented:
Still need to test this in stage.

dustymabe commented on docs/config.yaml:
```diff
@@ -55,6 +55,9 @@ streams:
       cosa_img: quay.io/coreos-assembler/coreos-assembler:stable
+      # OPTIONAL/TEMPORARY: override default coreos-assembler image to use for
+      # the controller pod in a multi-arch build
+      cosa_controller_img: quay.io/coreos-assembler/coreos-assembler:latest
```
suggestion: `cosa_build_arch_job_controller_img_hack`

This maybe should have a more complete comment with more context too.

dustymabe commented on jobs/build-arch.Jenkinsfile:
```diff
@@ -70,6 +70,8 @@ def cosa_img = params.COREOS_ASSEMBLER_IMAGE ?: pipeutils.get_cosa_img(pipecfg,
 
 def stream_info = pipecfg.streams[params.STREAM]
 
+def cosa_controller_img = stream_info.cosa_controller_img ?: cosa_img
```
this probably needs an appropriate comment about the context like the comment from the code block we're removing.

jlebon replied:
I was thinking the comment block deleted here would be moved (and reworded) to RHCOS' `config.yaml` above `cosa_build_arch_job_controller_img_hack` since that's the only place we'd actually configure it. WDYT?

dustymabe replied:
I guess that's sufficient. In that case let's add a comment in the docs/config.yaml about RHCOS being the only consumer of this right now and we'll be able to remove it in the future.

dustymabe reviewed (commented):
Thanks for working on this!

For all of the `bump-lockfile` and `kola-*` jobs we've never run them in a context other than latest cosa. Not sure if it makes sense to templatize them on stream or not like the build jobs.

jlebon commented on docs/config.yaml:
```diff
@@ -21,6 +21,10 @@ default_artifacts:
     - vmware
 
+# OPTIONAL: coreos-assembler image to build with. supports ${STREAM} variable.
+# If unset, defaults to `coreos-assembler:main` (ImageStream).
```
Ahh yup, that's leftover from when I initially wanted to do these changes in a different order (adding a knob before nuking the ImageStream).

jlebon commented:
> For all of the `bump-lockfile` and `kola-*` jobs we've never run them in a context other than latest cosa. Not sure if it makes sense to templatize them on stream or not like the build jobs. If we do decided it matters then we should make the automated spawning of the `kola-*` jobs from the `build*` jobs specify the COSA image as a parameter too.

I think for `bump-lockfile` it makes sense to respect the stream-level cosa image. Once we start having lockfiles in RHCOS, we'll want `bump-lockfile` to build with the same cosa it would've in e.g. `build`.

For the `kola-*` jobs it's less clear. It makes sense too for correctness, but with almost all the tests being external tests now, it's not as crucial to match. I'm also not sure if the cloud provisioning code in older cosa branches would need updating. So maybe we can leave those off for now and revisit if we actually hit something.

Will update.

jlebon commented:
Updated for comments! I've restored the `kola-*` jobs for now and split out the `bump-lockfile` change in a separate commit since it's technically a semantic change.

dustymabe commented on docs/config.yaml:
```diff
@@ -49,6 +53,13 @@ streams:
+      cosa_img: quay.io/coreos-assembler/coreos-assembler:stable
+      # OPTIONAL/TEMPORARY: Override default coreos-assembler image to use for
+      # the controller pod in the build-arch job. This is only useful when
+      # building with a cosa too old to support `cosa remote`, which should only
```
I think it's `remote-session`:

```suggestion
      # building with a cosa too old to support `cosa remote-session`, which should only
```

dustymabe commented on jobs/build-arch.Jenkinsfile:
```diff
@@ -66,8 +66,12 @@ properties([
     durabilityHint('PERFORMANCE_OPTIMIZED')
 ])
 
+def cosa_img = params.COREOS_ASSEMBLER_IMAGE ?: pipeutils.get_cosa_img(pipecfg, params.STREAM)
```
would this work? 

```suggestion
def cosa_img = params.COREOS_ASSEMBLER_IMAGE ?: \
                   pipeutils.get_cosa_img(pipecfg, params.STREAM)
```

These lines are getting a little long

jlebon replied:
I'm not sure, but I reworked it some other other into two lines and added a clarifying comment.

dustymabe reviewed (approved):
LGTM

jlebon commented:
Now tested that the various knobs in `build` and `build-arch` in stg correctly get wired!

jlebon commented on jobs/build-arch.Jenkinsfile:
```diff
@@ -46,7 +46,7 @@ properties([
       string(name: 'COREOS_ASSEMBLER_IMAGE',
              description: 'Override coreos-assembler image to use',
-             defaultValue: "coreos-assembler:main",
+             defaultValue: "",
```
One weird thing I noticed while testing in stg is that for some reason changing the default value back to the empty string didn't take hold. Even after running the job multiple times, it would still be `coreos-assembler:main`. I ended up having to rerun the `seed` job with `FORCE=1`.

dustymabe reviewed (approved):
LGTM
